### PR TITLE
feat: support modules to extend nitro builder

### LINKED ISSUE #1788: Nitro modules to extend core
### Describe the feature

Hello,

I'm currently asking myself how to share code between Nitro app.

Imagine, I'm building a server with auth on app A. I implement auth within code source A. Then, I'm building app B. I also need auth. How could I share code?

Currently, if I use Nuxt with Nitro, I can build a Nuxt module and modify directly the Nitro config to import some utils. If I use Nitro alone, it's not possible. In the same time, some features currently build for Nuxt could be only for Nitro and shared to other Nitro based framework. 

I could create a simple JS package but I will not be able to provide great DX like auto-import, to update the config, inject routes or more...

What do you think of modules for Nitro like in Nuxt?

### Additional information

- [ ] Would you be willing to help implement this feature?

### PR DESCRIPTION
**Pull request by Barbapapazes**
### 🔗 Linked issue

fix #1788

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR introduce the concept of modules like in Nuxt.

- Support for function in the `nitro.config` file
- Support for function in the `modules` directory

This PR is a first implementation. depending of the usage, we will update the modules feature.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: playground/nitro.config.ts [hunk 1]
```diff
@@ -1 +1,14 @@
-export default defineNitroConfig({});
+import { defineEventHandler } from "h3";
+
+export default defineNitroConfig({
+  modules: [
+    (_, nitro) => {
```
Comment (pi0) [inline playground/nitro.config.ts:5]: Having resolved options first for Nuxt was mostly a backyard compatibility. For nitro we can simply accept `(nitro)`

File: src/module/define.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+export function defineNitroModule(definition: ModuleDefinition | NitroModule) {
+  if (typeof definition === 'function') { return defineNitroModule({ setup: definition }) }
+
+  const module = defu(definition, { meta: {} })
+  const options = {}
+
+  async function normalizedModule(this: any, inlineOptions: any, nitro: Nitro = useNitro()) {
+    const res = await module.setup?.call(null as any, options, nitro) ?? {}
+
+    if (res === false) {
```
Comment (pi0) [inline src/module/define.ts:14]: This is not required for nitro

File: src/module/define.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import { ModuleDefinition, Nitro, NitroModule } from "../types";
+import { useNitro } from "../context";
+
+export function defineNitroModule(definition: ModuleDefinition | NitroModule) {
+  if (typeof definition === 'function') { return defineNitroModule({ setup: definition }) }
+
+  const module = defu(definition, { meta: {} })
+  const options = {}
+
+  async function normalizedModule(this: any, inlineOptions: any, nitro: Nitro = useNitro()) {
```
Comment (pi0) [inline src/module/define.ts:11]: this and inline options is not required for nitro

File: src/module/define.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { defu } from "defu";
+import { ModuleDefinition, Nitro, NitroModule } from "../types";
```
Comment (pi0) [inline src/module/define.ts:2]:
```suggestion
import type { ModuleDefinition, Nitro, NitroModule } from "../types";
```

File: src/module/install.ts
Comment (pi0) [inline src/module/install.ts:1]: Install API is not required for nitro. For Nuxt we had to introduce it

File: src/context.ts
Comment (pi0) [inline src/context.ts:1]: We might consider supporting multi instances for Nitro (also it is actually a case for prerenderer instance). Lets not introduce it

**Comment by pi0** [comment]
Thanks for picking up on it and nice you could take inspiration from the Nuxt module.

If you would like to work on your PR to be landed, i have added few comments to minify changes. Many of APIs and consideration in Nuxt happened over time and some based on legacy requirements we don't need to Nitro...

**Comment by Barbapapazes** [comment]
> Thanks for picking up on it and nice you could take inspiration from the Nuxt module.
> 
> If you would like to work on your PR to be landed, i have added few comments to minify changes. Many of APIs and consideration in Nuxt happened over time and some based on legacy requirements we don't need to Nitro...

Thanks! 🪄

I update this PR a little and allow modules to be specify in a `modules` folder. Before continuing, I would like to know if I should specify how end-user will write a module. In Nuxt, there is the object `meta` and the `nuxt-module-builder`. Should we have the same thing? How could we improve the module implementation to have a better usage?

**Comment by pi0** [comment]
I expect as simple as this:

```ts
import type { NitroModule } from 'nitropack'
export default <NitroModule>{ async setup(nitro) {} }
```

Let's not keep ahead of ourselves. This feature is initially intended for advanced use cases not user-facing as Nuxt modules are until we can figure out use-cases :)

**Comment by Barbapapazes** [comment]
> I expect as simple as this:
> 
> ```ts
> import type { NitroModule } from 'nitropack'
> export default <NitroModule>{ async setup(nitro) {} }
> ```
> 
> Let's not keep ahead of ourselves. This feature is initially intended for advanced use cases not user-facing as Nuxt modules are until we can figure out use-cases :)

So I think this PR is ready for a review!

File: src/build.ts [hunk 5]
```diff
@@ -504,6 +504,8 @@ async function _watch(nitro: Nitro, rollupConfig: RollupConfig) {
     join(dir, "api"),
     join(dir, "routes"),
     join(dir, "middleware", GLOB_SCAN_PATTERN),
+    join(dir, "plugins"),
```
Comment (pi0) [inline src/build.ts:507]: Please split unrelated changes to new PR

File: playground/nitro.config.ts [hunk 6]
```diff
@@ -1 +1,14 @@
-export default defineNitroConfig({});
+import { defineEventHandler } from "h3";
+
+export default defineNitroConfig({
```
Comment (pi0) [inline playground/nitro.config.ts:3]: Please consider not pushing commits for playground changes. They are for local testing only

File: tsconfig.json [hunk 7]
```diff
@@ -14,12 +17,28 @@
     "jsxFactory": "h",
     "jsxFragmentFactory": "Fragment",
     "paths": {
-      "nitropack": ["./src/index"],
-      "nitropack/config": ["./src/config"],
-      "#internal/nitro": ["./src/runtime/index"],
-      "#internal/nitro/*": ["./src/runtime/*"]
+      "nitropack": [
+        "./src/index"
```
Comment (pi0) [inline tsconfig.json:21]: Please consider moving unrelated changes like this + playground types to separate PRs to avoid blocking merges and allow more clear changelogs 🙏🏼

File: src/build.ts (hunk 5 above)
Comment (Barbapapazes) [reply src/build.ts:507]: Yes, you're right! https://github.com/unjs/nitro/pull/1800

File: src/module.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+export async function resolveNitroModule(
+  mod: NitroModuleInput
+): Promise<NitroModule> {
+  if (typeof mod === "string") {
+    // @ts-ignore
+    globalThis.defineNitroModule =
+      // @ts-ignore
+      globalThis.defineNitroModule || defineNitroModule;
+    mod = (await import(mod).then((m) => m.default)) as NitroModule;
```
Comment (pi0) [inline src/module.ts:15]: @Barbapapazes FYI, this works in nitro repo because in stup mode, jiti adds typescript support. We need explicit jiti if we wanted to support ts.

**Review by pi0 — APPROVED** [review]
